pwiz_tools/Shared/PanoramaClient/PanoramaUtil.cs

```diff
@@ -675,37 +349,128 @@ private static string GetErrorMessage(FolderState state, string folderPath, stri
-            return AppendErrorAndUri(stateError, error, requestUri);
+        public PanoramaServerException(string message, Exception e) : base(message, e)
+        {
+        }
+    }
+
+    public class ErrorMessageBuilder
+    {
+        private readonly string _error;
+        private string _errorDetail;
+        private string _exceptionMessage;
+        private LabKeyError _labkeyError;
+        private Uri _uri;
+        private string _responseString;
+
+        public ErrorMessageBuilder (string error)
+        {
+            _error = error;
+        }
+        public ErrorMessageBuilder ErrorDetail(string errorDetail)
+        {
+            _errorDetail = errorDetail;
+            return this;
+        }
+        public ErrorMessageBuilder LabKeyError(LabKeyError labkeyError)
+        {
+            _labkeyError = labkeyError;
+            return this;
         }
 
-        private static string AppendErrorAndUri(string mainMessage, string error, Uri uri)
+        public ErrorMessageBuilder Exception(Exception ex)
         {
-            var message = mainMessage;
+            Exception(ex, null);
+            return this;
+        }
 
-            if (error != null || uri != null)
+        // If the given exception is a WebException, this method tries to parse the response as JSON and
+        // get a LabKey error if it was set. The method accepts a function to parse the response in the 
+        // WebException so that tests can pass in a custom function.
+        public ErrorMessageBuilder Exception(Exception ex, Func<WebException, LabKeyError> labkeyErrorGetter)
+        {
+            _exceptionMessage = ex.Message;
+            var exception = ex as WebException;
+            if (exception != null && labkeyErrorGetter != null)
             {
-                var sb = new StringBuilder();
+                _labkeyError = labkeyErrorGetter(exception);
+            }
 
-                if (error != null)
-                {
-                    sb.AppendLine(string.Format(Resources.GenericState_AppendErrorAndUri_Error___0_, error));
-                }
+            return this;
+        }
 
-                if (uri != null)
-                {
-                    sb.AppendLine(string.Format(Resources.GenericState_AppendErrorAndUri_URL___0_, uri));
-                }
+        public ErrorMessageBuilder Uri(Uri requestUri)
+        {
+            _uri = requestUri;
+            return this;
+        }
 
-                message = TextUtil.LineSeparate(message, string.Empty, sb.ToString());
+        public ErrorMessageBuilder Response(JObject json)
+        {
+            if (json != null)
+            {
+                _responseString = json.ToString(Formatting.Indented);
             }
+            return this;
+        }
+        public ErrorMessageBuilder Response(string response)
+        {
+            _responseString = response;
+            return this;
+        }
 
-            return message;
+        public string Build()

```

Comment:
I would probably call this method "ToString" because that's what the method on "StringBuilder" is called, but "Build" is a fine name too.